File: src/lib/api/externalCalendars.ts
import { invoke } from "@tauri-apps/api/core";

/// External calendars are a read only overlay on the Sessions calendar, never
/// entities. See `src-tauri/src/external_calendars/mod.rs`.
export type CalendarProvider = "google" | "icloud";

export interface RemoteCalendar {
  id: string;
  name: string;
  color: string | null;
  selected: boolean;
}

export interface CalendarConnection {
  provider: CalendarProvider;
  account: string;
  calendars: RemoteCalendar[];
  lastSyncedAt: string | null;
  lastError: string | null;
}

export interface ExternalCalendarStatus {
  googleAvailable: boolean;
  connections: CalendarConnection[];
}

/// Timed events carry UTC ISO instants; all day events carry `YYYY-MM-DD` dates
/// with an exclusive `end`.
export interface ExternalEvent {
  id: string;
  provider: CalendarProvider;
  calendarId: string;
  calendarName: string;
  color: string | null;
  title: string;
  location: string | null;
  allDay: boolean;
  start: string;
  end: string;
}

export function externalCalendarStatus(): Promise<ExternalCalendarStatus> {
  return invoke("external_calendar_status");
}

/// Opens the browser for Google's sign in and resolves once it completes.
export function connectGoogleCalendar(): Promise<CalendarConnection> {
  return invoke("connect_google_calendar");
}

export function cancelGoogleCalendarConnect(): Promise<void> {
  return invoke("cancel_google_calendar_connect");
}

export function connectIcloudCalendar(
  appleId: string,
  appPassword: string,
): Promise<CalendarConnection> {
  return invoke("connect_icloud_calendar", { appleId, appPassword });
}

export function setExternalCalendarSelected(
  provider: CalendarProvider,
  calendarId: string,
  selected: boolean,
): Promise<CalendarConnection> {
  return invoke("set_external_calendar_selected", { provider, calendarId, selected });
}

export function disconnectExternalCalendar(provider: CalendarProvider): Promise<void> {
  return invoke("disconnect_external_calendar", { provider });
}

/// Fetches from every connected provider; failures are recorded per connection.
export function syncExternalCalendars(): Promise<CalendarConnection[]> {
  return invoke("sync_external_calendars");
}

/// Cached events of selected calendars from `from` through `to` (`YYYY-MM-DD`).
export function listExternalEvents(from: string, to: string): Promise<ExternalEvent[]> {
  return invoke("list_external_events", { from, to });
}

File: apps/desktop/src-tauri/src/commands/external_calendars.rs
use crate::error::AppResult;
use crate::external_calendars::{
    google, icloud, Connection, ExternalCalendarState, ExternalEvent, Provider,
};
use serde::Serialize;
use tauri::State;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ExternalCalendarStatus {
    /// False when this build has no Google OAuth client configured.
    google_available: bool,
    connections: Vec<Connection>,
}

#[tauri::command]
pub async fn external_calendar_status(
    state: State<'_, ExternalCalendarState>,
) -> AppResult<ExternalCalendarStatus> {
    Ok(ExternalCalendarStatus {
        google_available: google::available(),
        connections: state.connections().await,
    })
}

#[tauri::command]
pub async fn connect_google_calendar(
    app: tauri::AppHandle,
    state: State<'_, ExternalCalendarState>,
) -> AppResult<Connection> {
    google::connect(&state, &app).await
}

#[tauri::command]
pub fn cancel_google_calendar_connect(state: State<'_, ExternalCalendarState>) {
    google::cancel_connect(&state);
}

#[tauri::command]
pub async fn connect_icloud_calendar(
    state: State<'_, ExternalCalendarState>,
    apple_id: String,
    app_password: String,
) -> AppResult<Connection> {
    icloud::connect(&state, &apple_id, &app_password).await
}

#[tauri::command]
pub async fn set_external_calendar_selected(
    state: State<'_, ExternalCalendarState>,
    provider: Provider,
    calendar_id: String,
    selected: bool,
) -> AppResult<Connection> {
    state.set_selected(provider, &calendar_id, selected).await
}

#[tauri::command]
pub async fn disconnect_external_calendar(
    state: State<'_, ExternalCalendarState>,
    provider: Provider,
) -> AppResult<()> {
    state.disconnect(provider).await
}

/// Never fails as a whole: each connection records its own error and keeps
/// its cached events.
#[tauri::command]
pub async fn sync_external_calendars(
    state: State<'_, ExternalCalendarState>,
) -> AppResult<Vec<Connection>> {
    Ok(state.sync().await)
}

/// Cached events only, so the calendar never waits on the network.
#[tauri::command]
pub async fn list_external_events(
    state: State<'_, ExternalCalendarState>,
    from: String,
    to: String,
) -> AppResult<Vec<ExternalEvent>> {
    Ok(state.events_between(&from, &to).await)
}
